Removing unused parameters fixes #3486

juanignaciosl opened the pull request:
@rafatower CR this fix for #3486, please.

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3449/))

rafatower commented on app/controllers/admin/visualizations_controller.rb:
```diff
@@ -570,12 +570,12 @@ def get_visualization_and_table(table_id, schema, filter)
 
-  def get_visualization_and_table_from_table_id(table_id, user, filter)
-    user_table = Carto::UserTable.where({ id: table_id, user_id: user.id }).first
+  def get_visualization_and_table_from_table_id(table_id)
```
Isn't there a privacy issue here? what if I request by table_id something that does not belong to me?

juanignaciosl replied:
Good question. Privacy concerns are covered at controller actions ([take a look at `show`, for example](https://github.com/CartoDB/cartodb/blob/3486-NoMethodError_undefined_method_id_for_nil_NilClass_at_get_visualization_and_table_from_table_id/app/controllers/admin/visualizations_controller.rb#L67)). There's some duplication and it should be refactored, but IMHO refactoring old controllers is far beyond scope.

rafatower commented on app/controllers/admin/visualizations_controller.rb:
```diff
@@ -570,12 +570,12 @@ def get_visualization_and_table(table_id, schema, filter)
-  def get_visualization_and_table_from_table_id(table_id, user, filter)
-    user_table = Carto::UserTable.where({ id: table_id, user_id: user.id }).first
+  def get_visualization_and_table_from_table_id(table_id)
+    user_table = Carto::UserTable.where({ id: table_id }).first
```
The scope of this is a bugfix. This introduces a bigger problem than it fixes. The user_id shall not be removed from the equation.

juanignaciosl replied:
No, no, it does not. Privacy concerns are always checked afterwards.

The bug it fixes is user fetching with organizations involved. Since [user depends on subdomain](https://github.com/CartoDB/cartodb/blob/3486-NoMethodError_undefined_method_id_for_nil_NilClass_at_get_visualization_and_table_from_table_id/app/controllers/admin/visualizations_controller.rb#L569) organization public maps don't fetch user right. Maybe @Kartones can shine some light on this?

Kartones replied:
yes, user depends on subdomain, but my worry is that we should check privacy on all this actions:
```
before_filter :resolve_visualization_and_table, only: [:show, :public_table, :public_map,
                                                         :show_organization_public_map, :show_organization_embed_map,
                                                         :show_protected_public_map, :show_protected_embed_map,
                                                         :embed_map ]
```

rafatower commented:
@Kartones an extra pair of eyes can help here

Kartones commented:
Uff, I'd not change this unless fully covered by tests that explicitly try to do security-sensitive actions (like trying to request tables from other user being logged in) on all actions that call `get_visualization_and_table`.

It worries me too because although many methods have their own security checks, we must be sure all of them do check properly security to avoid security leaks. Removing the `and user_id=xxxx` filter feels dangerous without refactoring the controllers code and adding the mentioned tests.

So I'd say... a) postpone, b) add specific tests or c) small refactor of controller to make sure all calls have some before_filter(s) (and probably still some tests :P ).

juanignaciosl commented:
Ok, since I have many PRs open right now I'll hold this for a while (a) until I can do (b) and (c), hopefully before public release.

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3478/))

juanignaciosl commented on app/controllers/admin/visualizations_controller.rb:
```diff
@@ -73,14 +73,11 @@ def show
     unless @visualization.has_permission?(current_user, Visualization::Member::PERMISSION_READWRITE)
       if request.original_fullpath =~ %r{/tables/}
-        CartoDB.url(self, 'public_table_map', {id: request.params[:id], redirected:true})
+        return redirect_to CartoDB.url(self, 'public_table_map', {id: request.params[:id], redirected:true})
```
@Kartones could you please take a look at this change? I think this (old) code missed actual redirection and did nothing, really.

Kartones replied:
uh, yes, or maybe in the url rewrite I ate the `return redirect_to`...

juanignaciosl commented on app/controllers/admin/visualizations_controller.rb:
```diff
@@ -415,7 +401,7 @@ def track_embed
   # @param visualization CartoDB::Visualization::Member
   def disallowed_type?(visualization)
-    return false if visualization.nil?
+    return true if visualization.nil?
```
I'm interpreting `nil` as a `disallowed_type` (which might not be a common Ruby idiom but it's literally the way nils are interpreted in serious languages :P). Nevertheless it's consistent with current code, as far as I've read, if you take a look at the refactor that centralises the 404 on the filter.

juanignaciosl commented:
@Kartones , @rafatower , take a second look at this, please. As you can see I've added some tests (which actually disclosed some errors, both in old and new code). I've added a couple of comments about the diff as well.

Kartones commented on app/controllers/admin/visualizations_controller.rb:
```diff
@@ -90,8 +87,7 @@ def show
 
   def public_table
-    return(pretty_404) if @visualization.nil? || @visualization.private?
-    return(pretty_404) if disallowed_type?(@visualization)
```
Careful, you've removed the disallowed_type check in many views... slides are not viewable themselves, they only exist as children of a parent visualization, so that check must exist in public table, public map, etc.

juanignaciosl replied:
Yes, I've removed it from the actions and moved it to `resolve_visualization_and_table` (https://github.com/CartoDB/cartodb/pull/3493/files#diff-af16d8490cbdf7991cb9f204b72b1d14R443), which is run as a `before_filter`. This way 404 behaviour is consistent everytime you are supposed to fetch the visualization and it's not found, without code duplication.

Kartones commented on app/models/user.rb:
```diff
@@ -1090,6 +1089,8 @@ def db_size_in_bytes
       retry unless attempts > 1
       Rollbar.report_exception(e)
+      # INFO: we need to return something to avoid 'disabled' return value
+      nil
```
What about returning `ERROR #{e.message}` or something similar? Not so needed, but to keep "unsilencing" this errors when there's no rollbar

juanignaciosl replied:
That way types are inconsistent. Returning nil you can easily do a fallback, but checking returning type is nonsense. I was returning `nil` because of that, and because of being consistent with previous, non-failing, behaviour. IMHO there're only two ways to proceed here: failing, or returning nil. I'd prefer failing, maybe with an increasing timeout on retry, but returning `nil` is _slightly_ more consistent with previous behaviour. What do you think, @rafatower ? Having such a key method as not failure safe is a huge PITA :(

Kartones replied:
You're right... then better the nil, but we should start wrapping all rollbar calls (using the existing [error notifier I mean](https://github.com/CartoDB/cartodb/blob/master/config/initializers/error_notifier.rb)) so in dev/test they output stuff and we don't silence the errors and by mistake return "disabled"s...

Kartones commented:
+1 when you add back the checks of disallowed_type, as else we will have problems with Odyssey slide type

Kartones commented on app/controllers/admin/visualizations_controller.rb:
```diff
@@ -454,6 +440,7 @@ def resolve_visualization_and_table
     if @visualization && @visualization.user
       @more_visualizations = more_visualizations(@visualization.user, @visualization)
     end
+    render_pretty_404 if disallowed_type?(@visualization)
```
ahh sorry, didn't saw it!

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3488/))

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3501/))

rafatower commented:
:+1:

juanignaciosl commented on app/controllers/admin/visualizations_controller.rb:
```diff
@@ -291,13 +283,12 @@ def show_organization_public_map
 
     respond_to do |format|
-      format.html { render 'public_map', layout: false }
+      format.html { render 'public_map' }
```
Without this change the request is ok (200) but content is empty, weird :-\ cc @Kartones @rafatower

juanignaciosl replied:
@rochoa this line is yours (months ago), do you recall why layout was false? @Kartones just told me to ask @xavijam , so I'm summoning both just in case any of you know :-)

xavijam replied:
If forces to not set any layout, avoiding problems in case there was a 
```ruby
layout 'application'
```
at the beginning of the class.

rochoa replied:
thx @xavijam for clarifying it :-)

juanignaciosl commented:
@iriberri I need QA for this. It's deployed in staging, in ded01. We want to make the new visualization endpoint public, and this PR is related (although it doesn't need any feature flag enabled. Could you do some testing about the following?
- Public page (map/table).
- Embed maps.

It should be done both for organization users and normal users. Negative tests (checking permissions under organization and private/protected both inside and outside) are also important for security.

Please let me know when you begin testing this, and stop and notify me as soon as anything looks weird.

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3518/))

iriberri commented:
Ask me tomorrow to test this :+1:

iriberri commented:
Found this one but it's in production https://github.com/CartoDB/cartodb/issues/3555
